### PR TITLE
New Integration - Perch

### PR DESCRIPTION
**Pull request by amshamah419**
## Status
Ready

## Related Issues
fixes: demisto/etc#16598

## Description
New Integration for Perch Security.

## Must have
- [x] Tests
- [ ] Documentation (with link to it)
- [x] Code Review

### CODE REVIEW

File: Integrations/Perch/Perch.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+
+USERNAME = demisto.params().get('credentials').get('identifier')
+PASSWORD = demisto.params().get('credentials').get('password')
+API_KEY = demisto.params().get('api-key')
+FETCH_TIME = int(demisto.params().get('fetch_time', '7'))
+SERVER = demisto.params()['url'][:-1] if (demisto.params()['url'] and demisto.params()['url'].endswith('/')) else \
+    demisto.params()['url']
+USE_SSL = not demisto.params().get('insecure', False)
+BASE_URL = SERVER + '/v1'
+if not demisto.params().get('proxy'):
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please remove and use the `handle_proxy` function

File: Integrations/Perch/Perch.yml [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+      description: ID given by perch of the incident.
+      type: String
+    - contextPath: Perch.Indicator.CreatedBy
+      description: ID of incident's creator.
+      type: Number
+  dockerimage: demisto/python3:3.7.2.214
+  isfetch: true
+  runonce: false
+  script: ''
+  type: python
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.yml]: please add python3 subtype  see https://github.com/demisto/content/blob/9a08201d3246ec6cef0034472e10160d3a9f6041/Scripts/CherwellIncidentUnlinkTask/CherwellIncidentUnlinkTask.yml
for example

File: Integrations/Perch/Perch.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+            last_run = {'time': 1561017202}
+            fetch_alerts(last_run, headers)
+        demisto.results('ok')
+    except Exception as err:
+        return_error(str(err))
+
+
+''' COMMANDS MANAGER / SWITCH PANEL '''
+
+# LOG('Command being called is %s' % (demisto.command()))
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: remove comment

File: Integrations/Perch/Perch.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+        )
+        if res.status_code == 403:
+            return_error('Connection forbidden. Please verify your API key is valid.')
+        elif res.status_code not in {200, 201}:
+            return_error('Error in API call to Perch Integration [%d] - %s' % (res.status_code, res.reason))
+
+    except requests.exceptions.ConnectionError as error:
+        return_error("Failed to establish a new connection: {error}".format(error=type(error)))
+
+    return res.json()
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please wrap the .json with a try and return error like "failed to parse response to error" and print the response contents.

File: Integrations/Perch/Perch.py [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+            fetch_alerts(last_run, headers)
+        demisto.results('ok')
+    except Exception as err:
+        return_error(str(err))
+
+
+''' COMMANDS MANAGER / SWITCH PANEL '''
+
+# LOG('Command being called is %s' % (demisto.command()))
+demisto.info('Command being called is {}'.format(demisto.command()))
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please use py3 formatting like f'my string{my_var}'

File: Integrations/Perch/Perch.py [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+            data=data,
+            headers=headers
+        )
+        if res.status_code == 403:
+            return_error('Connection forbidden. Please verify your API key is valid.')
+        elif res.status_code not in {200, 201}:
+            return_error('Error in API call to Perch Integration [%d] - %s' % (res.status_code, res.reason))
+
+    except requests.exceptions.ConnectionError as error:
+        return_error("Failed to establish a new connection: {error}".format(error=type(error)))
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please use py3 formatting like f'my string{my_var}'

File: Integrations/Perch/Perch.py [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+    hr = ''
+    ec = {
+        "Perch": {
+            "Alert": []
+        }
+    }  # type: dict
+    for alert in res_results:
+        alert_hr, alert_ec = format_alerts(alert)
+        ec['Perch']['Alert'].append(alert_ec)
+        hr += tableToMarkdown('{title}'.format(title=alert_ec.get('Title')), alert_hr)
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please use py3 formating

File: Integrations/Perch/Perch.py [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+            'EntryContext': ec
+        })
+
+
+def get_community_command():
+    headers = authenticate()
+    args = demisto.args()
+    params = alerts_params(args)
+    id = args.get('id')
+    url = '/communities/{id}'.format(id=id)
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please use py3 formatting

File: Integrations/Perch/Perch.py [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+    res = http_request('POST', url, headers=headers, data=data)
+    indicator_hr, indicator_ec = format_indicator(res[0])
+    hr = ''
+    ec = {
+        "Perch": {
+            "Indicator": []
+        }
+    }  # type: dict
+    ec['Perch']['Indicator'].append(indicator_ec)
+    hr += tableToMarkdown('{title}'.format(title=indicator_hr.get('Title')), indicator_hr)
```
Comment (yuvalbenshalom) [inline Integrations/Perch/Perch.py]: please use py3 formattin

**Comment by amshamah419** [comment]
@kirbles19 These are the files that need review.
https://github.com/demisto/content/blob/perch/Integrations/Perch/Perch.yml
https://github.com/demisto/content/blob/perch/Integrations/Perch/perch-description.md

**Comment by kirbles19** [comment]
All files reviewed.

**Pull request merged**